File: csharp/demo/DllPathResolver.cs
using System;
using System.Runtime.InteropServices;
using System.Reflection;
using System.Collections.Generic;
using System.IO;

public class DllPathResolver
{
    static List<string> gAssemblyResolvePaths = new List<string>();
    static ResolveEventHandler gResolveHandler;

    static System.Reflection.Assembly OnAssemblyResolveFail(object sender, ResolveEventArgs args)
    {
        string dllName = args.Name.Substring(0, args.Name.IndexOf(",")) + ".dll";

        foreach (string resolvePath in gAssemblyResolvePaths)
        {
            string  assemblyPath = Path.Combine(resolvePath, dllName);
            if (File.Exists(assemblyPath))
            {
                System.Console.WriteLine("DllPathResolver: Loading assembly from path: " + assemblyPath);
                System.Reflection.Assembly tempAssembly = System.Reflection.Assembly.LoadFrom(assemblyPath);
                return tempAssembly;
            }
        }
        return null;
    }

    public static void AddPath(string dllPath)
    {
        dllPath = Path.GetFullPath(dllPath);
        if (!dllPath.EndsWith("\\"))
        {
            dllPath += "\\";
        }
        gAssemblyResolvePaths.Add(dllPath);
    }

    public static void Resolve(string[] args)
    {
        // First, locate the root of the distribution, and use that to construct
        // paths for locating native DLL's.

        string rootDir = FindDistributionRootDir();

        string binRelativePath;
        if (IntPtr.Size == 4)
        {
            binRelativePath = "csharp/bin/x86";
        }
        else
        {
            binRelativePath = "csharp/bin/x64";
        }
        string rcfProtoBinDir = Path.Combine(rootDir, binRelativePath);
        AddPath(rcfProtoBinDir);

        string rcfProtoAssemblyDir = Path.Combine(rootDir, "csharp/bin");
        AddPath(rcfProtoAssemblyDir);

        // Second, parse command line for any further paths.
        for (int i = 0; i < args.Length; i++)
        {
            if ((args[i] == "--dllpath" || args[i] == "-d") && i + 1 < args.Length)
            {
                string dllPath = args[i + 1];
                AddPath(dllPath);
            }
        }

        // Configure the assembly resolve handler (used to find .NET assemblies).
        gResolveHandler = new ResolveEventHandler(OnAssemblyResolveFail);
        AppDomain.CurrentDomain.AssemblyResolve += gResolveHandler;

        // Configure the PATH environment variable (used to find native DLL's).
        string envPath = Environment.GetEnvironmentVariable("PATH");
        foreach (string dllPath in gAssemblyResolvePaths)
        {
            envPath = envPath + ";" + dllPath;
        }
        Environment.SetEnvironmentVariable("PATH", envPath);
    }

    public static string FindDistributionRootDir()
    {
        string codeBase = Assembly.GetExecutingAssembly().CodeBase;
        UriBuilder uri = new UriBuilder(codeBase);
        string path = Uri.UnescapeDataString(uri.Path);
        path = Path.GetDirectoryName(path);

        string testDir = path;
        string prevTestDir = path;
        while (testDir != null && Directory.Exists(testDir))
        {
            string testDllPath = Path.Combine(testDir, "csharp\\bin\\RCFProto_NET.dll");
            if (File.Exists(testDllPath))
            {
                break;
            }

            // Go up one level.
            testDir = Path.GetFullPath(Path.Combine(testDir, ".."));
            if (testDir.Equals(prevTestDir))
            {
                testDir = null;
            }

            prevTestDir = testDir;
        }

        return testDir;
    }
}

File: java/demo/DllPathResolver.java

import java.io.File;
import java.io.IOException;

import java.lang.reflect.*;

public class DllPathResolver
{
    public static void resolve(String[] args) throws Exception
    {
        String rootDir = findDistributionRootDir();
        
        String javaLibPath = System.getProperty( "java.library.path" );
        String runningOnOs = System.getProperty("os.name");

        // Class separator symbol differs between Windows and Unix.
        String cpSeparator = ":";
        if (runningOnOs.startsWith("Windows"))
        {
            cpSeparator = ";";
        }
        
        String arch = System.getProperty("sun.arch.data.model");
        String binRelativePath = "";
        if (arch.equals("32"))
        {
            binRelativePath = "java/bin/x86";
        }
        else
        {
            binRelativePath = "java/bin/x64";
        }
        
        // Add default bin location in the distribution, to javaLibPath.
        File binPath = new File(rootDir, binRelativePath);
        javaLibPath = javaLibPath + cpSeparator + binPath;      

        // Add any --dllpath command line entries, to javaLibPath.
        for (int i=0; i<args.length; i++)
        {
            if (    (args[i].equals("--dllpath") || args[i].equals("-d"))
                &&  i+1<args.length)
            {
                File dllDir = new File(args[i+1]);
                String extraPath = dllDir.getCanonicalPath();
                javaLibPath = javaLibPath + cpSeparator + extraPath;
            }
        }

        // This hack enables us to modify java.lib.path at runtime.
        // http://blog.cedarsoft.com/2010/11/setting-java-library-path-programmatically/
        // java.lib.path is used by the JVM when loading the JNI dll.
        System.setProperty( "java.library.path", javaLibPath );
        Field fieldSysPath = ClassLoader.class.getDeclaredField( "sys_paths" );
        fieldSysPath.setAccessible( true );
        fieldSysPath.set( null, null );

    }
    
    public static String findDistributionRootDir() throws IOException
    {
        String runningOnOs = System.getProperty("os.name");
        String javaLibPath = System.getProperty("java.library.path");

        String dllFullName = "";

        if (runningOnOs.startsWith("Windows"))
        {
            dllFullName = "RCFProto_Java_impl.dll";
        }
        else
        {
            dllFullName = "libRCFProto_Java_impl.so";
        }
        
        String arch = System.getProperty("sun.arch.data.model");
        String binRelativePath = "";
        if (arch.equals("32"))
        {
            binRelativePath = "java/bin/x86/" + dllFullName;
        }
        else
        {
            binRelativePath = "java/bin/x64/" + dllFullName;
        }
        
        // Search upwards until we find the native DLL. 
        String initPath = new File(".").getCanonicalPath();
        File testDir = new File(initPath);
        File prevTestDir = new File(initPath);
        while (testDir != null && testDir.exists())
        {
            File testDllPath = new File(testDir, binRelativePath);
            if (testDllPath.exists())
            {
                break;
            }

            // Go up one level.
            testDir = new File(testDir, "..");
            testDir = new File(testDir.getCanonicalPath());
            if (testDir.equals(prevTestDir))
            {
                testDir = null;
            }

            prevTestDir = testDir;
        }
        
        return testDir.getCanonicalPath();

    }
    
}
